src/qengine/state/opencl.cpp

```diff
@@ -10,80 +10,90 @@
 // See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
 // for details.
 
+#include <memory>
+
 #include "oclengine.hpp"
 #include "qengine_opencl.hpp"
 
 namespace Qrack {
 
 #define CMPLX_NORM_LEN 5
 
-void QEngineOCL::InitOCL()
+void QEngineOCL::SetDevice(const int& dID)
 {
-    clObj = OCLEngine::Instance();
+    if ((dID >= clObj->GetDeviceCount()) || (dID < -1)) {
+        throw "Invalid OpenCL device selection";
+    } else {
+        deviceID = -1;
+    }
+    device_context = clObj->GetDeviceContextPtr(deviceID);
+    context = device_context->context;
+    queue = device_context->queue;
+}
 
-    queue = *(clObj->GetQueuePtr());
-    cl::Context context = *(clObj->GetContextPtr());
+void QEngineOCL::InitOCL(int devID)
+{
+    clObj = OCLEngine::Instance();
```

Comment:
Also, clObj seems to be unused; if it’s there just to make sure OCLEngine has been initialized first, then add a comment to that effect.

Reply:
See comment below; removed.